wallet-core/src/eip2333.rs

```diff
@@ -0,0 +1,631 @@
+    lamport_combined[..HKDF_DIGESTS]
+        .clone_from_slice(&lamport_0[..HKDF_DIGESTS]);
+    lamport_combined[HKDF_DIGESTS..HKDF_DIGESTS * 2]
+        .clone_from_slice(&lamport_1[..HKDF_DIGESTS]);
+
+    // for i  in 1, .., 255
+    //    lamport_PK = lamport_PK | SHA256(lamport_0[i])
+    // for i  in 1, .., 255
+    //    lamport_PK = lamport_PK | SHA256(lamport_1[i])
+    let mut lamport_pk = [0u8; HKDF_OUTPUT_SIZE * 2];
+    for i in 0..HKDF_DIGESTS * 2 {
+        let sha_slice = &Sha256::digest(lamport_combined[i]);
+        lamport_pk[i * SHA256_DIGEST_SIZE..(i + 1) * SHA256_DIGEST_SIZE]
+            .clone_from_slice(sha_slice);
+    }
+
+    // compressed_lamport_PK = SHA256(lamport_PK)
+    Sha256::digest(lamport_pk).to_vec()
+}
+
+/// Derives a BLS scalar using HKDF extraction and expansion.
+///
+/// # EIP2333 Specification:
+///
+/// Inputs
+///  - `IKM`, a secret octet string >= 256 bits in length
+///  - `key_info`, an optional octet string (default="", the empty string)
+///
+/// Output
+///  - `SK`, the corresponding secret key, an integer 0 <= SK < r.
+///
+/// Definitions
+///  - `L`: integer given by `ceil((3 * ceil(log2(r)))` / 16).(L=48)
+///  - `r`: the order of the BLS 12-381 curve defined in the v4 draft IETF BLS
+///    signature scheme standard
+///
+/// Procedure
+///
+///  1. `salt = "BLS-SIG-KEYGEN-SALT-"`
+///  2. `SK = 0`
+///  3. `while SK == 0:`
+///  4. `salt = H(salt)`
+///  5. `PRK = HKDF-Extract(salt, IKM || I2OSP(0, 1))`
+///  6. `OKM = HKDF-Expand(PRK, key_info || I2OSP(L, 2), L)`
+///  7. `SK = OS2IP(OKM) mod r`
+///  8. `return SK`
+///
+///
+/// # Arguments
+/// * `ikm`      - The initial keying material for HKDF.
+/// * `key_info` - The info for HKDF.
+///
+/// # Returns
+/// The derived secret key.
+///
+/// # Panics
+/// Panics if HKDF extraction/expansion fail.
+#[allow(clippy::similar_names)]
+fn hkdf_mod_r(ikm: &[u8], key_info: &[u8]) -> BlsScalar {
+    // IKM || I2OSP(0, 1)
+    let ikm_combined = [ikm, &[0u8]].concat();
+    // key_info || I2OSP(L, 2)
+    let key_info_combined = [key_info, &[0u8, 48u8]].concat();
+
+    // HKDF output size L (=48)
+    let mut okm: [u8; 48] = [0u8; 48];
+
+    // SK = 0
+    let mut sk = BlsScalar::zero();
+
+    // salt = "BLS-SIG-KEYGEN-SALT-"
+    // salt = H(salt)
+    let mut salt = Sha256::digest(b"BLS-SIG-KEYGEN-SALT-");
+
+    while sk.is_zero().into() {
+        // PRK = HKDF-Extract(salt, IKM || I2OSP(0, 1))
+        // OKM = HKDF-Expand(PRK, key_info || I2OSP(L, 2), L)
+        hkdf(&salt, ikm_combined.as_ref(), &key_info_combined, &mut okm);
+
+        // Convert okm to a 64-byte little-endian value
+        let mut okm_le_64 = [0u8; 64];
+        okm.reverse();
+        okm_le_64[..48].copy_from_slice(&okm);
+
+        // SK = OS2IP(OKM) mod r
+        sk = BlsScalar::from_bytes_wide(&okm_le_64);
+
+        // salt = H(salt)
+        if sk.is_zero().into() {
+            salt = Sha256::digest(salt);
+        }
```

Comment:
I don't understand why this if-statement is here

Reply:
To avoid computing the digest if the next iteration is not going to be executed.
Logically this digest should be computed at the beginning of each iteration, but it was a bit complicated to do it because the initialization value of `salt` has a different length.

Reply:
My bad, I didn't see that the if-statement is part of the `while` loop.